modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/operation/DeferredExecutorCallback.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.module.extension.internal.runtime.operation;
+
+import static java.lang.Thread.currentThread;
+import static org.mule.runtime.core.api.util.ClassUtils.setContextClassLoader;
+import org.mule.runtime.extension.api.runtime.operation.CompletableComponentExecutor.ExecutorCallback;
+
+/**
+ * Class used to defer an executor callback until this is closed, so the execution can be intercepted without calling the executor
+ * callback methods inside the interception.
+ *
+ * <code>
+ *   ExecutorCallback delegate;
+ *
+ *   try (DeferredExecutorCallback deferredCallback = new DeferredExecutorCallback(delegate)) {
+ *     withSomeInterceptionLogic(() -> {
+ *       deferredCallback.complete(someValue);
+ *       // At this point, the delegate.complete(someValue) hasn't been called yet.
+ *     }
+ *   }
+ *
+ *   // When the execution goes out of the try, delegate.complete(someValue) is called. It's useful because the
+ *   // interception logic isn't applied here.
+ * </code>
+ *
+ * The methods of this class are synchronized because the callback completion can be invoked from other thread than the close
+ * method. This shouldn't introduce a performance degradation because they're only called from two threads, and only once.
+ */
+class DeferredExecutorCallback implements ExecutorCallback, AutoCloseable {
```

Comment:
Just a comment: I think it's ok to leave this a package class (preferable to an inner class) as this will be used only in this package.

Reply:
Yes, the idea was adding JavaDocs without increasing the length of the DefaultExecutorMediator code